docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,526 @@
+runtime will query for the referenced resource and only when the specified field
+is found in referenced resource, the reconciliation loop will progress. Otherwise
+the request will get requeued for future processing.
+
+* Watchers for the referenced resource type will not be implemented in the first
+revision. The first version of this implementation will mostly be used for finding
+identifier of referenced resource and any update in referenced resource should not
+change it's identifier. As mentioned earlier, deletion will also be handled in
+coming revisions.
+
+* The code to determine when the referenced resource is synced will be generated,
+and additional hooks will be added for customization. Ex: Do not just check presence
+of a field but also check if the Status is available etc...
+
+### Low Level Implementation
+
+#### generator.yaml Changes
+
+```go
+type FieldConfig struct {
+    ...
+    Reference *ReferenceConfig `json:"references,omitempty"`
+    ...
+}
+
+// ReferenceConfig contains the instructions for how to add the referenced resource
+// configuration for a field.
+type ReferenceConfig struct {
+	// Kind mentions the K8s Kind of referenced resource
+	Kind string `json:"kind"`
+	// FieldPath refers to the the path of field which should be copied
+	// from referenced resource into TargetFieldName
+	FieldPath string `json:"field_path"`
+	// TargetFieldName is spec field name which gets value from
+	// referenced resource's FieldPath
+	TargetFieldName string `json:"target_field"`
+	// IsList mentions whether the new field should be AWSResourceReference or
+	// List of AWSResourceReference
+	IsList          bool   `json:"is_list"`
+	// Required mentions whether either of AWSResourceReference or
+	// TargetFieldName must be present in custom resource
+	Required        bool   `json:"required"`
+}
+```
+
+`APIGatewayv2 generator.yaml` Example:
+```yaml
+resources:
+  Integration:
+    fields:
+      Api:
+        references:
+          kind: API
+          field_path: Status.APIID
+          target_field: APIID
+          is_list: false
+          required: true
+      ApiId:
+        is_required: false
+```
+NOTE: `ApiId` field is marked as not required because the new reference field
+`Api` can also provide ApiId.
+
+#### New AWSResourceReference Type
+
+```go
+type AWSResourceReference struct {
+	Name *string `json:"name,omitempty"`
+}
+```
+
+and the new manifests will look like following:
+
+a) Same as earlier (keeping backwards compatibility)
+
+```yaml
+apiVersion: apigatewayv2.services.k8s.aws/v1alpha1
+kind: Integration
+metadata:
+  name: my-integration
+spec:
+  apiID: my-api-id
+  integrationType: HTTP_PROXY
+  integrationURI: https://example.org
+  integrationMethod: GET
+  payloadFormatVersion: "1.0"
+```
+
+b) Using the API resource reference
+```yaml
+apiVersion: apigatewayv2.services.k8s.aws/v1alpha1
+kind: Integration
+metadata:
+  name: my-integration
+spec:
+  api:
+    name: my-api
```

Comment:
Does this proposal assume that the resources should be in the same namespace? will cross namespace reference be supported?

Reply:
Yes, that is correct. There is no plan to support cross namespace references. 
TL;DR: In K8s namespaces are security boundaries and we should not be referring resources across namespace. There is more discussion on the same topic in PR #856